File: core/src/com/unciv/logic/automation/unit/religious_unit_automation.py
from typing import Optional, List

from com.unciv import Constants
from com.unciv.logic.automation import Automation, ThreatLevel
from com.unciv.logic.city import City
from com.unciv.logic.civilization.diplomacy import DiplomacyFlags, RelationshipLevel
from com.unciv.logic.map.mapunit import MapUnit
from com.unciv.models import UnitActionType
from com.unciv.models.ruleset.unique import UniqueType
from com.unciv.ui.screens.worldscreen.unit.actions import UnitActions
from com.unciv.logic.automation.unit import UnitAutomation

class ReligiousUnitAutomation:
    """Handles AI automation for religious units."""

    @staticmethod
    def automate_missionary(unit: MapUnit) -> None:
        """Automate missionary unit actions.
        
        Args:
            unit: The missionary unit to automate
        """
        if (unit.religion != unit.civ.religion_manager.religion?.name 
            or unit.religion is None):
            return unit.disband()

        our_cities_without_religion = [
            city for city in unit.civ.cities
            if city.religion.get_majority_religion() != unit.civ.religion_manager.religion
        ]
        
        def is_valid_spread_religion_target(city: City) -> bool:
            """Check if a city is a valid target for spreading religion.
            
            Args:
                city: The city to check
                
            Returns:
                bool: Whether the city is a valid target
            """
            diplomacy_manager = unit.civ.get_diplomacy_manager(city.civ)
            if diplomacy_manager?.has_flag(DiplomacyFlags.AgreedToNotSpreadReligion):
                # See NextTurnAutomation - these are the conditions under which AI agrees to religious demands
                # If they still hold, keep the agreement, otherwise we can renege
                if diplomacy_manager.relationship_level() == RelationshipLevel.Ally:
                    return False
                if Automation.threat_assessment(unit.civ, city.civ) >= ThreatLevel.High:
                    return False
            return True

        def rank_city_for_religion_spread(city: City) -> int:
            """Rank a city for religion spread priority.
            
            Args:
                city: The city to rank
                
            Returns:
                int: Lower value means higher priority
            """
            rank = city.get_center_tile().aerial_distance_to(unit.get_tile())
            
            diplomacy_manager = unit.civ.get_diplomacy_manager(city.civ)
            if diplomacy_manager?.has_flag(DiplomacyFlags.AgreedToNotSpreadReligion):
                rank += 10  # Greatly discourage, but if the other options are too far away we'll take it anyway
                
            return rank
        
        city = None
        if our_cities_without_religion:
            city = min(
                our_cities_without_religion,
                key=lambda city: city.get_center_tile().aerial_distance_to(unit.get_tile())
            )
        else:
            valid_cities = [
                city for city in unit.civ.game_info.get_cities()
                if (city.religion.get_majority_religion() != unit.civ.religion_manager.religion
                    and city.civ.knows(unit.civ)
                    and not city.civ.is_at_war_with(unit.civ)
                    and not city.religion.is_protected_by_inquisitor(unit.religion)
                    and is_valid_spread_religion_target(city))
            ]
            if valid_cities:
                city = min(valid_cities, key=rank_city_for_religion_spread)

        if not city:
            return

        destination = next(
            (tile for tile in sorted(
                city.get_tiles(),
                key=lambda tile: tile.aerial_distance_to(unit.get_tile())
            )
            if (unit.movement.can_move_to(tile) or tile == unit.get_tile())
            and unit.movement.can_reach(tile)),
            None
        )
        
        if not destination:
            return

        unit.movement.head_towards(destination)

        if (unit.get_tile() in city.get_tiles() 
            and unit.civ.religion_manager.may_spread_religion_now(unit)):
            UnitActions.invoke_unit_action(unit, UnitActionType.SpreadReligion)

    @staticmethod
    def automate_inquisitor(unit: MapUnit) -> None:
        """Automate inquisitor unit actions.
        
        Args:
            unit: The inquisitor unit to automate
        """
        civ_religion = unit.civ.religion_manager.religion

        if (unit.religion != civ_religion?.name 
            or unit.religion is None):
            return unit.disband()  # No need to keep a unit we can't use, as it only blocks religion spreads of religions other that its own

        holy_city = unit.civ.religion_manager.get_holy_city()
        city_to_convert = ReligiousUnitAutomation._determine_best_inquisitor_city_to_convert(unit)
        pressure_deficit = (
            city_to_convert.religion.get_pressure_deficit(civ_religion?.name)
            if city_to_convert else 0
        )

        cities_to_protect = [
            city for city in unit.civ.cities
            if (city.religion.get_majority_religion() == civ_religion
                # We only look at cities that are not currently protected or are protected by us
                and (not city.religion.is_protected_by_inquisitor()
                     or unit.get_tile() in city.get_center_tile().get_tiles_in_distance(1)))
        ]

        # Cities with most populations will be prioritized by the AI
        city_to_protect = max(
            cities_to_protect,
            key=lambda city: city.population.population,
            default=None
        )

        destination_city = None
        if (city_to_convert
            and (city_to_convert == holy_city
                 or pressure_deficit > Constants.ai_prefer_inquisitor_over_missionary_pressure_difference
                 or (city_to_convert.religion.is_blocked_holy_city 
                     and city_to_convert.religion.religion_this_is_the_holy_city_of == civ_religion?.name))
            and unit.has_unique(UniqueType.CanRemoveHeresy)):
            destination_city = city_to_convert
        elif (city_to_protect 
              and unit.has_unique(UniqueType.PreventSpreadingReligion)):
            if holy_city and not holy_city.religion.is_protected_by_inquisitor():
                destination_city = holy_city
            else:
                destination_city = city_to_protect
        elif city_to_convert:
            destination_city = city_to_convert

        if not destination_city:
            return

        destination_tile = next(
            (tile for tile in sorted(
                destination_city.get_center_tile().neighbors,
                key=lambda tile: tile.aerial_distance_to(unit.current_tile)
            )
            if (unit.movement.can_move_to(tile) or tile == unit.get_tile())
            and unit.movement.can_reach(tile)),
            None
        )
        
        if not destination_tile:
            return

        unit.movement.head_towards(destination_tile)

        if (city_to_convert 
            and unit.get_tile().get_city() == destination_city):
            UnitActions.invoke_unit_action(unit, UnitActionType.RemoveHeresy)

    @staticmethod
    def _determine_best_inquisitor_city_to_convert(unit: MapUnit) -> Optional[City]:
        """Determine the best city for an inquisitor to convert.
        
        Args:
            unit: The inquisitor unit
            
        Returns:
            Optional[City]: The best city to convert, or None if no valid targets
        """
        if (unit.religion != unit.civ.religion_manager.religion?.name 
            or not unit.has_unique(UniqueType.CanRemoveHeresy)):
            return None

        holy_city = unit.civ.religion_manager.get_holy_city()
        if (holy_city 
            and holy_city.religion.get_majority_religion() != unit.civ.religion_manager.religion):
            return holy_city

        blocked_holy_city = next(
            (city for city in unit.civ.cities
             if (city.religion.is_blocked_holy_city 
                 and city.religion.religion_this_is_the_holy_city_of == unit.religion)),
            None
        )
        if blocked_holy_city:
            return blocked_holy_city

        return max(
            (city for city in unit.civ.cities
             if (city.religion.get_majority_religion()
                 and city.religion.get_majority_religion() != unit.civ.religion_manager.religion
                 and city.get_center_tile().aerial_distance_to(unit.current_tile) <= 20)),
            key=lambda city: city.religion.get_pressure_deficit(
                unit.civ.religion_manager.religion?.name
            ),
            default=None
        )

    @staticmethod
    def found_religion(unit: MapUnit) -> None:
        """Automate founding a religion.
        
        Args:
            unit: The unit to use for founding religion
        """
        city_to_found_religion_at = None
        if (unit.get_tile().is_city_center() 
            and not unit.get_tile().owning_city.is_holy_city()):
            city_to_found_religion_at = unit.get_tile().owning_city
        else:
            city_to_found_religion_at = next(
                (city for city in unit.civ.cities
                 if (not city.is_holy_city()
                     and unit.movement.can_move_to(city.get_center_tile())
                     and unit.movement.can_reach(city.get_center_tile()))),
                None
            )

        if not city_to_found_religion_at:
            return

        if unit.get_tile() != city_to_found_religion_at.get_center_tile():
            unit.movement.head_towards(city_to_found_religion_at.get_center_tile())
            return

        UnitActions.invoke_unit_action(unit, UnitActionType.FoundReligion)

    @staticmethod
    def enhance_religion(unit: MapUnit) -> None:
        """Automate enhancing a religion.
        
        Args:
            unit: The unit to use for enhancing religion
        """
        # Try go to a nearby city
        if not unit.get_tile().is_city_center():
            UnitAutomation.try_enter_own_closest_city(unit)

        # If we were unable to go there this turn, unable to do anything else
        if not unit.get_tile().is_city_center():
            return

        UnitActions.invoke_unit_action(unit, UnitActionType.EnhanceReligion) 

File: rust_port/src/automation/unit/religious_unit_automation.rs
use crate::models::civilization::diplomacy::{DiplomacyFlags, RelationshipLevel};
use crate::models::map::mapunit::MapUnit;
use crate::models::city::City;
use crate::models::UnitActionType;
use crate::models::ruleset::unique::UniqueType;
use crate::automation::Automation;
use crate::automation::ThreatLevel;
use crate::ui::screens::worldscreen::unit::actions::UnitActions;
use crate::constants::AI_PREFER_INQUISITOR_OVER_MISSIONARY_PRESSURE_DIFFERENCE;

/// Contains logic for automating religious unit actions
pub struct ReligiousUnitAutomation;

impl ReligiousUnitAutomation {
    /// Automates missionary movement and actions
    pub fn automate_missionary(unit: &mut MapUnit) {
        if unit.religion != unit.civ.religion_manager.religion.as_ref().map(|r| r.name.clone())
            || unit.religion.is_none() {
            return unit.disband();
        }

        let our_cities_without_religion: Vec<&City> = unit.civ.cities.iter()
            .filter(|city| city.religion.get_majority_religion() != unit.civ.religion_manager.religion)
            .collect();

        /// Returns whether a city is a valid target for spreading religion
        fn is_valid_spread_religion_target(unit: &MapUnit, city: &City) -> bool {
            let diplomacy_manager = unit.civ.get_diplomacy_manager(city.civ);
            if let Some(manager) = diplomacy_manager {
                if manager.has_flag(DiplomacyFlags::AgreedToNotSpreadReligion) {
                    // See NextTurnAutomation - these are the conditions under which AI agrees to religious demands
                    // If they still hold, keep the agreement, otherwise we can renege
                    if manager.relationship_level() == RelationshipLevel::Ally {
                        return false;
                    }
                    if Automation::threat_assessment(unit.civ, city.civ) >= ThreatLevel::High {
                        return false;
                    }
                }
            }
            true
        }

        /// Returns a rank for city priority in religion spreading (lower is better)
        fn rank_city_for_religion_spread(unit: &MapUnit, city: &City) -> i32 {
            let mut rank = city.get_center_tile().aerial_distance_to(unit.get_tile());

            let diplomacy_manager = unit.civ.get_diplomacy_manager(city.civ);
            if let Some(manager) = diplomacy_manager {
                if manager.has_flag(DiplomacyFlags::AgreedToNotSpreadReligion) {
                    rank += 10; // Greatly discourage, but if the other options are too far away we'll take it anyway
                }
            }

            rank
        }

        let city = if !our_cities_without_religion.is_empty() {
            our_cities_without_religion.into_iter()
                .min_by_key(|city| city.get_center_tile().aerial_distance_to(unit.get_tile()))
        } else {
            unit.civ.game_info.get_cities()
                .iter()
                .filter(|city| city.religion.get_majority_religion() != unit.civ.religion_manager.religion)
                .filter(|city| city.civ.knows(unit.civ) && !city.civ.is_at_war_with(unit.civ))
                .filter(|city| !city.religion.is_protected_by_inquisitor(unit.religion.as_ref()))
                .filter(|city| is_valid_spread_religion_target(unit, city))
                .min_by_key(|city| rank_city_for_religion_spread(unit, city))
        };

        let city = match city {
            Some(city) => city,
            None => return,
        };

        let destination = city.get_tiles()
            .iter()
            .filter(|tile| unit.movement.can_move_to(tile) || *tile == unit.get_tile())
            .min_by_key(|tile| tile.aerial_distance_to(unit.get_tile()))
            .filter(|tile| unit.movement.can_reach(tile));

        let destination = match destination {
            Some(tile) => tile,
            None => return,
        };

        unit.movement.head_towards(destination);

        if city.get_tiles().contains(unit.get_tile()) && unit.civ.religion_manager.may_spread_religion_now(unit) {
            UnitActions::invoke_unit_action(unit, UnitActionType::SpreadReligion);
        }
    }

    /// Automates inquisitor movement and actions
    pub fn automate_inquisitor(unit: &mut MapUnit) {
        let civ_religion = unit.civ.religion_manager.religion.as_ref();

        if unit.religion != civ_religion.map(|r| r.name.clone()) || unit.religion.is_none() {
            // No need to keep a unit we can't use, as it only blocks religion spreads of religions other than its own
            return unit.disband();
        }

        let holy_city = unit.civ.religion_manager.get_holy_city();
        let city_to_convert = Self::determine_best_inquisitor_city_to_convert(unit);
        let pressure_deficit = city_to_convert
            .map(|city| city.religion.get_pressure_deficit(civ_religion.map(|r| r.name.as_str())))
            .unwrap_or(0);

        let cities_to_protect = unit.civ.cities.iter()
            .filter(|city| city.religion.get_majority_religion() == civ_religion)
            // We only look at cities that are not currently protected or are protected by us
            .filter(|city| {
                !city.religion.is_protected_by_inquisitor() ||
                unit.get_tile().is_in(city.get_center_tile().get_tiles_in_distance(1))
            });

        // Cities with most populations will be prioritized by the AI
        let city_to_protect = cities_to_protect
            .max_by_key(|city| city.population.population);

        let destination_city = match (city_to_convert, holy_city, city_to_protect) {
            (Some(convert), _, _)
                if (convert == holy_city.as_ref()
                    || pressure_deficit > AI_PREFER_INQUISITOR_OVER_MISSIONARY_PRESSURE_DIFFERENCE
                    || (convert.religion.is_blocked_holy_city
                        && convert.religion.religion_this_is_the_holy_city_of == civ_religion.map(|r| r.name.clone())))
                    && unit.has_unique(UniqueType::CanRemoveHeresy) => Some(convert),
            (_, _, Some(protect)) if unit.has_unique(UniqueType::PreventSpreadingReligion) => {
                if let Some(holy) = holy_city {
                    if !holy.religion.is_protected_by_inquisitor() {
                        Some(holy)
                    } else {
                        Some(protect)
                    }
                } else {
                    Some(protect)
                }
            },
            (Some(convert), _, _) => Some(convert),
            _ => None,
        };

        let destination_city = match destination_city {
            Some(city) => city,
            None => return,
        };

        let destination_tile = destination_city.get_center_tile().neighbors()
            .iter()
            .filter(|tile| unit.movement.can_move_to(tile) || *tile == unit.get_tile())
            .min_by_key(|tile| tile.aerial_distance_to(unit.current_tile))
            .filter(|tile| unit.movement.can_reach(tile));

        let destination_tile = match destination_tile {
            Some(tile) => tile,
            None => return,
        };

        unit.movement.head_towards(destination_tile);

        if city_to_convert.is_some() && unit.get_tile().get_city() == Some(destination_city) {
            UnitActions::invoke_unit_action(unit, UnitActionType::RemoveHeresy);
        }
    }

    /// Determines the best city for an inquisitor to convert
    fn determine_best_inquisitor_city_to_convert(unit: &MapUnit) -> Option<&City> {
        if unit.religion != unit.civ.religion_manager.religion.as_ref().map(|r| r.name.clone())
            || !unit.has_unique(UniqueType::CanRemoveHeresy) {
            return None;
        }

        let holy_city = unit.civ.religion_manager.get_holy_city();
        if let Some(holy_city) = holy_city {
            if holy_city.religion.get_majority_religion() != unit.civ.religion_manager.religion {
                return Some(holy_city);
            }
        }

        let blocked_holy_city = unit.civ.cities.iter()
            .find(|city| {
                city.religion.is_blocked_holy_city &&
                city.religion.religion_this_is_the_holy_city_of == unit.religion
            });

        if let Some(city) = blocked_holy_city {
            return Some(city);
        }

        unit.civ.cities.iter()
            .filter(|city| city.religion.get_majority_religion().is_some())
            .filter(|city| city.religion.get_majority_religion() != unit.civ.religion_manager.religion)
            // Don't go if it takes too long
            .filter(|city| city.get_center_tile().aerial_distance_to(unit.current_tile) <= 20)
            .max_by_key(|city| city.religion.get_pressure_deficit(
                unit.civ.religion_manager.religion.as_ref().map(|r| r.name.as_str())
            ))
    }

    /// Handles founding a religion with a great prophet
    pub fn found_religion(unit: &mut MapUnit) {
        let city_to_found_religion_at = if unit.get_tile().is_city_center()
            && !unit.get_tile().owning_city().unwrap().is_holy_city() {
            unit.get_tile().owning_city()
        } else {
            unit.civ.cities.iter()
                .find(|city| {
                    !city.is_holy_city()
                        && unit.movement.can_move_to(city.get_center_tile())
                        && unit.movement.can_reach(city.get_center_tile())
                })
        };

        let city_to_found_religion_at = match city_to_found_religion_at {
            Some(city) => city,
            None => return,
        };

        if unit.get_tile() != city_to_found_religion_at.get_center_tile() {
            unit.movement.head_towards(city_to_found_religion_at.get_center_tile());
            return;
        }

        UnitActions::invoke_unit_action(unit, UnitActionType::FoundReligion);
    }

    /// Handles enhancing a religion with a great prophet
    pub fn enhance_religion(unit: &mut MapUnit) {
        // Try go to a nearby city
        if !unit.get_tile().is_city_center() {
            UnitAutomation::try_enter_own_closest_city(unit);
        }

        // If we were unable to go there this turn, unable to do anything else
        if !unit.get_tile().is_city_center() {
            return;
        }

        UnitActions::invoke_unit_action(unit, UnitActionType::EnhanceReligion);
    }
}